lib/nimble_totp.ex

```diff
@@ -185,14 +222,51 @@ defmodule NimbleTOTP do
   In this example by validating first against the current time, but also against 30 seconds
   ago, we allow the _previous_ code, to be still valid.
   """
+  @spec valid?(binary(), [option() | validate_option()]) :: boolean()
   def valid?(secret, otp, opts \\ [])
 
   def valid?(secret, <<a1, a2, a3, a4, a5, a6>>, opts) do
     <<e1, e2, e3, e4, e5, e6>> = verification_code(secret, opts)
 
     (bxor(e1, a1) ||| bxor(e2, a2) ||| bxor(e3, a3) ||| bxor(e4, a4) ||| bxor(e5, a5) |||
-       bxor(e6, a6)) === 0
+       bxor(e6, a6)) === 0 and not reused?(opts)
   end
 
   def valid?(_secret, _otp, _opts), do: false
+
+  @spec reused?([option() | validate_option()]) :: boolean()
+  defp reused?(opts) do
+    if since = opts |> Keyword.get(:since) |> to_unix() do
+      time = opts |> Keyword.get(:time, System.os_time(:second)) |> to_unix()
+      period = Keyword.get(opts, :period, @default_totp_period)
+
+      Integer.floor_div(time, period) <= Integer.floor_div(since, period)
+    else
+      false
+    end
+  end
+
+  @spec to_unix(nil) :: nil
+  defp to_unix(nil), do: nil
+
+  defp to_unix(%NaiveDateTime{} = naive_date_time),
+    do: naive_date_time |> DateTime.from_naive!("Etc/UTC") |> to_unix()
```

Comment:
This doesn't have a spec of it's own, because it would overlap with `Calendar.datetime()`, but should be handled differently.